tools/codegen/gen.py

```diff
@@ -276,10 +276,21 @@ def func(f: NativeFunction) -> Optional[str]:
                     cuda_guard = """\
     // DeviceGuard omitted
 """
+            mem_overlap_check = ""
+            if (dispatch == "CPU" or dispatch == "CUDA") and ('backward' not in impl_name):
+                if f.func.name.name.inplace:
+                    mem_overlap_check = """\
+    at::assert_no_internal_overlap(self);
+"""
+                if f.func.is_out_fn():
+                    for out_argument in f.func.out_arguments:
+                        mem_overlap_check += f"""\
+    at::assert_no_internal_overlap({out_argument.name});
+"""
 
             return f"""\
 {returns_type} {name}({args_str}) {{
-{cuda_guard}{return_kw}{impl_name}({args_exprs_str});
+{mem_overlap_check}{cuda_guard}{return_kw}{impl_name}({args_exprs_str});
```

Comment:
If I am understanding this correctly, it looks like mem_overlap_check would be generated for CPUType and CUDAType. I think this is good and what we want: the mem overlap check doesn't apply to other backends.

Reply:
Let's check with some of the composability folks to make sure that it's OK to add things into the codegen. cc @ezyang, we wanted to auto-generate memory overlap checks directly into the CPUType / CUDAType codegen for in-place and out= operations. Does this sound OK?